src/memory-analyzer/analyze_symbol.cpp

```diff
@@ -0,0 +1,375 @@
+}
+
+/// Get memory snapshot as C code
+std::string gdb_value_extractort::get_snapshot_as_c_code()
+{
+  code_blockt generated_code;
+
+  allocate_objects.declare_created_symbols(generated_code);
+
+  for(auto const &pair : assignments)
+  {
+    generated_code.add(code_assignt(pair.first, pair.second));
+  }
+
+  return c_converter.convert(generated_code);
+}
+
+/// Get memory snapshot as symbol table
+symbol_tablet gdb_value_extractort::get_snapshot_as_symbol_table()
+{
+  symbol_tablet snapshot;
+
+  for(const auto &pair : assignments)
+  {
+    const symbol_exprt &symbol_expr = to_symbol_expr(pair.first);
+    const irep_idt id = symbol_expr.get_identifier();
+
+    INVARIANT(symbol_table.has_symbol(id), "symbol must exist in symbol table");
+
+    const symbolt &symbol = symbol_table.lookup_ref(id);
+
+    symbolt snapshot_symbol(symbol);
+    snapshot_symbol.value = pair.second;
+
+    snapshot.insert(snapshot_symbol);
+  }
+
+  // Also add type symbols to the snapshot
+  for(const auto &pair : symbol_table)
+  {
+    const symbolt &symbol = pair.second;
+
+    if(symbol.is_type)
+    {
+      snapshot.insert(symbol);
+    }
+  }
+
+  return snapshot;
+}
+
+void gdb_value_extractort::add_assignment(const exprt &lhs, const exprt &value)
+{
+  assignments.push_back(std::make_pair(lhs, value));
+}
+
+exprt gdb_value_extractort::get_char_pointer_value(
+  const exprt &expr,
+  const memory_addresst &memory_location,
+  const source_locationt &location)
+{
+  PRECONDITION(expr.type().id() == ID_pointer);
+  PRECONDITION(is_c_char_type(expr.type().subtype()));
+  PRECONDITION(!memory_location.is_null());
+
+  auto it = values.find(memory_location);
+
+  if(it == values.end())
+  {
+    std::string c_expr = c_converter.convert(expr);
+    pointer_valuet value = gdb_api.get_memory(c_expr);
+    CHECK_RETURN(value.string);
+    std::string string = *value.string;
+
+    string_constantt init(string);
+    CHECK_RETURN(to_array_type(init.type()).is_complete());
+
+    symbol_exprt dummy(pointer_typet(init.type(), config.ansi_c.pointer_width));
+    code_blockt assignments;
+
+    const symbol_exprt new_symbol =
+      to_symbol_expr(allocate_objects.allocate_automatic_local_object(
+        assignments, dummy, init.type()));
+
+    add_assignment(new_symbol, init);
+
+    values.insert(std::make_pair(memory_location, new_symbol));
+
+    return new_symbol;
+  }
+  else
+  {
+    return it->second;
+  }
+}
+
+exprt gdb_value_extractort::get_non_char_pointer_value(
+  const exprt &expr,
+  const memory_addresst &memory_location,
+  const source_locationt &location)
+{
+  PRECONDITION(expr.type().id() == ID_pointer);
+  PRECONDITION(!is_c_char_type(expr.type().subtype()));
+  PRECONDITION(!memory_location.is_null());
+
+  auto it = values.find(memory_location);
+
+  if(it == values.end())
+  {
+    values.insert(std::make_pair(memory_location, nil_exprt()));
+
+    const typet target_type = expr.type().subtype();
+
+    symbol_exprt dummy(expr.type());
```

Comment:
Missing name (this constructor actually is deprecated).

Reply:
Done.